Read The Internet Workshop Bounty (#1506)
hackclub/hackclub: workshops/read_the_internet/README.md

---
name: 'Reading the Internet'
description: 'Use python to read html and extract any information you can find!'
author: '@RafaelCenzano'
img: 'https://cloud-p5pdve0kf.vercel.app/2webscrape.jpg'
---

# Reading the Internet
We can read all the content on websites, but wouldn't it be cool if we could automatically read specific data from websites whenever you wanted? If this interested you, this workshop is for you. I will be showing you how to use webscraping in Python.
**To Note:** not all webpages can be read using this technique, if data is loaded in using ajax or after the inital page request this method may not work.

[Final Code](https://repl.it/@SavageCoder77/Read-The-Internet#main.py)
[Live Result](https://Read-The-Internet.savagecoder77.repl.run)

## Prerequisites

> You should have a basic understanding of **HTML** and **Python** or how to use variables and objects.

Reviewer: ```suggestion
You should have a basic understanding of **HTML** and **Python** and how to use variables and objects.
```

> ## Let's do some setup

Reviewer: ```suggestion
## Setup
```

![Lets start gif](https://cloud-p23nck0rj.vercel.app/1lets-start.gif)

> We will be using [repl.it](https://repl.it) an online code editor. This allows you to write and use this code from any computer!

Reviewer: ```suggestion
We will be using [repl.it](https://repl.it), an online code editor that allows you to write and run code from any computer!
```

We will start by creating a Python 3 repl (You can easily create one by clicking this link https://repl.it/languages/python3).

We now need to install a couple of external libraries. Follow the image below to find where to add the packages. Once in this menu you will need to install `beautifulsoup4`, `requests`, and `lxml`.  Documentation on the packages: [BS4](https://www.crummy.com/software/BeautifulSoup/bs4/doc/), [requests](https://requests.readthedocs.io/en/master/), and [LXML](https://lxml.de).
![Repl.it packages menu](https://cloud-p5pdve0kf.vercel.app/0ezgif.com-video-to-gif.gif)

> #### A short description on the packages

Reviewer: ```suggestion
#### About the packages
```

---
name: 'Reading the Internet'
description: 'Use Python to read HTML and extract any information you can find!'
author: '@RafaelCenzano'
img: 'https://cloud-p5pdve0kf.vercel.app/2webscrape.jpg'
---

> We can read all the content on websites, but wouldn't it be cool if we could automatically read specific data from websites whenever you wanted? If this interested you, this workshop is for you. I will be showing you how to use webscraping in Python.

Reviewer 2: This doesn't do the workshop justice. It's not just a tutorial on webscraping, you're actually building a tool that scrapes COVID case numbers from a website and displays them as text. This paragraph should be edited to more clearly show what you're going to build.

*Note: not all webpages can be read using this technique. If data is loaded in using ajax or after the inital page request this method may not work.*

[Final Result and Code](https://repl.it/@SavageCoder77/Read-The-Internet#main.py)

> 

Reviewer 2: Maybe also include a screenshot of the covid case numbers displayed in the terminal

## Prerequisites
You should have a basic understanding of **HTML** and **Python** and how to use variables and objects.

## Setup
![Lets do it gif](https://cloud-p23nck0rj.vercel.app/1lets-start.gif)

We will be using [repl.it](https://repl.it), an online code editor that allows you to write and run code from any computer!

Spin up a Python3 repl by [clicking here](https://repl.it/languages/python3).

Before we can start coding, we need to install a couple of external libraries. Follow the image below to find where to add the packages. Once in this menu you will need to install `beautifulsoup4`, `requests`, and `lxml`.

![Repl.it packages menu](https://cloud-p5pdve0kf.vercel.app/0ezgif.com-video-to-gif.gif)

#### What each package does

> `lxml` is a C package used by beautiful soup that parses the HTML, `beautifulsoup4` is for working with the parsed HTML, and `requests` is for requesting the webpage/html.

Reviewer 2: This is a good simple explanation for how each library applies to this workshop, but I think you should also include a little more information about what each library is. I still don't fully understand exactly what BeautifulSoup is—"working with the parsed HTML" doesn't mean much to me. Similar to `requests`: requests is a library that makes it super easy to make HTTP requests in Python, and in this workshop we're going to use it to "request" the HTML of a website.

You should make it super super clear, almost overexplaining what each library does, because understanding this is essential to understanding the rest of the workshop, and in its current state most club members will likely get lost if they don't understand what these libraries do.

<details>

<summary> Learn more about the packages: </summary>

- [BS4](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [requests](https://requests.readthedocs.io/en/master/)
- [LXML](https://lxml.de) 

</details>

## Lets start coding!

![Start coding gif](https://cloud-5bmb4t5tp.vercel.app/1start-coding.gif)

Once you've installed the packages, open the `main.py` file in the sidebar on the left. Then, import the packages you just installed.

```python
from bs4 import BeautifulSoup as bs
import requests
```

Now we can start webscraping! We will begin by requesting our webpage and extracting the HTML from the request object.

```python
url = 'https://ncov2019.live/data'
# request the website
r = requests.get(url)
# get the HTML from the request
page = r.text
```

Next we will create the BeautifulSoup object and use `lxml` to parse the HTML.

```python
soup = bs(page, 'lxml')
```

Now that we have our BeautifulSoup object we can go all of the HTML! 

### Working with the parsed HTML

![HTML gif](https://cloud-5bmb4t5tp.vercel.app/2html.gif)

In the next step, we'll be working with data from an HTML table. So, before, we continue, a quick mini-lesson on HTML tables:

> 

Reviewer 2: I think in this step it should be a lot more clear why you're suddenly learning about HTML tables. When I first read this it was a little jarring because I was like "wait what? i thought we were writing python. Why are we suddenly pausing to learn about HTML tables?"

Here's an example of an HTML table:

```html
<table>
    <tr>
        <td>Country</td>
        <td>Cases</td>
        <td>Column</td>
        <td>Column 2</td>
    </tr>
    <tr>
        <td>United States</td>
        <td>1,000,000</td>
        <td>Column data</td>
        <td>Column 2 data</td>
    </tr>
    <tr>
        <td>Canada</td>
        <td>100,000</td>
        <td>Column data</td>
        <td>Column 2 data</td>
    </tr>
</table>
```

A HTML table can be seen as a 2 dimensional array or list. The table contains rows (`tr` tags). Each row contains columns (`td` tags). Within the `td` tags is the data or text we can webscrape.

### Python time!

> We're going to use BeautifulSoup to find an HTML table element. We want to find a specific table by ID `sortable_table_world` as seen in the picture. The code below finds the first table that matches that ID by using the `find` function and specifying the id as an attribute. The `attrs` is an optional argument and is a dictionary that can be set to any id, class, or HTML identifier.

Reviewer 2: During my first few reads, it wasn't totally clear to me what exactly this meant. I kinda get it now, but I'm still having trouble picturing why we're getting data from an HTML table or why this is necessary for webscraping, and I'm having trouble picturing what this step is actually doing. I think it'll be a lot more confusing to club members who are doing this workshop.

![nCOVID table open with inspect element](https://cloud-p5pdve0kf.vercel.app/1inspect_element.png)

> 

Reviewer 2: You just finished explaining HTML tables, so you should have a transition to the next step here, before the code block.

```python
# get world data table
table = soup.find('table', attrs={'id':'sortable_table_world'})
```

> After this code runs we now have a new beautiful soup object called `table` that contains the table and child elements within that table.

Reviewer 2: Throughout the workshop, you keep referring the library as `beautiful soup`, but it's not a soup that's beautiful, it's a library called `BeautifulSoup`. I think you should go through and change all instances of `beautiful soup` to `BeautifulSoup`

> With `table` we will use the `findAll` function to get all the `tr` tags from the table. `tableRows` is a Python list with a beautiful soup object for each `tr` tag found in the table.

Reviewer 2: what is the `findAll` function? Where does it come from? Should be explained

```python
# get table rows
tableRows = table.findAll('tr')
```

> Now that we have all the rows in a Python list, we can iterate through the list and try to find the United States or you can input any Country you want.

Reviewer 2: awkward wording, this can be shortened or made a little clearer

> Before getting to the looping part though we will need to take a look at some of our outputs. This may happen to you when webscraping, the text has weird formatting. In these next steps I'm going to show the text from the website with the weird formatting and how we can use Python to remove that.

Reviewer 2: this paragraph is also kinda awkward wording and i don't feel like i'm fully understanding/picturing what i should be understanding/picturing after i read it. should be made clearer

> I will just use the second row and first column for this example

Reviewer 2: now i'm lost. second row and first column of what?

```python
tableCols = tableRows[1].findAll('td')
tableCol = tableCols[0]
tableCol.text
```
This returns `'\r\n          TOTAL\r\n        '` with weird spacing and newline and indents. With Python we can use the `strip` function on the string and we will end with this code `tableCol.text.strip()` and this text `'TOTAL'`. Now that we got that handled lets get to reading the data!

> First we have a have a for each loop to loop through every row in the table. You make notice the `[1:]`, this was added to make the loop just a little more effcient because the first row at index `[0]` is just the table header with no data. Next we use an if statment to check if the current row's country is the United States (This can be replaced by any country). **Do Note** that we use the `.find()` function because we are only checking the first column and `.text[2:]` because the website has a weird astricks we need to remove. Once we find the country of our choice, we use `findAll()` to get all the columns in the row and print out some data.

Reviewer 2: This is a very long paragraph with multiple explanations and instructions. It's very hard to follow. should be broken up

```python
for row in tableRows[1:]:
    if 'United States' == row.find('td').text[2:].strip():
        columns = row.findAll('td')
        print(columns[0].text[2:].strip())
        print(f'Confirmed Cases: {columns[1].text.strip()}')
        print(f'New Cases: {columns[3].text.strip()}')
        print(f'Critical Cases: {columns[5].text.strip()}')
```
The output at the time of writing this workshop is:
```
United States
Confirmed Cases: 9,566,690
New Cases: 28,845
Critical Cases: 88,127
```
Scary Numbers! Hopefully at the time of reading this workshop things won't be too much worse :(

> Here is some example code for checking for multiple countries

Reviewer 2: this confuses me because it's not worded as an instruction, so i'm not sure if i should copy/paste this code in order to continue the workshop or what exactly i should be doing

```python
countries = ['United States', 'Spain', 'Canada']
for row in tableRows[1:]:
    if row.find('td').text[2:].strip() in countries:
        columns = row.findAll('td')
        print(columns[0].text[2:].strip())
        print(f'Confirmed Cases: {columns[1].text.strip()}')
        print(f'New Cases: {columns[3].text.strip()}')
        print(f'Critical Cases: {columns[5].text.strip()}\n')
```

> ### Thats it! You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!

Reviewer 2: This is way too long for a heading. It shouldn't be a full sentence. Should probably be something like:

```
### That's it!

You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!
```

> ### Thats it! You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!

Reviewer 2: Also this reads almost like a runon. May be useful to add some kind of punctuation after "COVID19 data"

> ![Sponge bob done gif](https://cloud-5bmb4t5tp.vercel.app/0completed.gif)

Reviewer 2: Click on "Preview" in this GitHub file and scroll down here. You may notice that the text appears next to the image instead of under it. This is because single new lines in markdown aren't interpreted as new lines. This is a problem I've been noticing throughout the workshop. It looks weird at first, but the proper formatting should be:

```
### That's it! You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!

![Sponge bob done gif](url)

following paragraph
```

That way it'll look correct on the website

> You can check out more advanced/in depth projects here. In this project you will be shown how to use selenium to webscrape that will allow you to webscrape pages like amazon or google that use ajax to load as you scroll (https://www.analyticsvidhya.com/blog/2020/08/web-scraping-selenium-with-python/). Check out this workshop by realpython that goes more in depth and links to other resources that can allow you to webscrape more complicated websites (https://realpython.com/beautiful-soup-web-scraper-python/).

Reviewer 2: This should not be dumped all in one paragraph one sentence after another. Instead, you should start a new section called "Hacking" and inside the section include each of these things in a list
